bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeperAdmin.java

```diff
@@ -1128,122 +1118,40 @@ public void processResult(int rc, String s, Object ctx) {
-            // Format all ledger metadata layout
-            bkc.ledgerManagerFactory.format(conf, zkc);
-
-            // Clear underreplicated ledgers
-            try {
-                ZKUtil.deleteRecursive(zkc, ZkLedgerUnderreplicationManager.getBasePath(conf.getZkLedgersRootPath())
-                        + BookKeeperConstants.DEFAULT_ZK_LEDGERS_ROOT_PATH);
-            } catch (KeeperException.NoNodeException e) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("underreplicated ledgers root path node not exists in zookeeper to delete");
-                }
-            }
-
-            // Clear underreplicatedledger locks
-            try {
-                ZKUtil.deleteRecursive(zkc, ZkLedgerUnderreplicationManager.getBasePath(conf.getZkLedgersRootPath())
-                        + '/' + BookKeeperConstants.UNDER_REPLICATION_LOCK);
-            } catch (KeeperException.NoNodeException e) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("underreplicatedledger locks node not exists in zookeeper to delete");
-                }
-            }
-
-            // Clear the cookies
-            try {
-                ZKUtil.deleteRecursive(zkc, conf.getZkLedgersRootPath()
-                        + "/cookies");
-            } catch (KeeperException.NoNodeException e) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("cookies node not exists in zookeeper to delete");
+                    doFormat = IOUtils
+                        .confirmPrompt("Ledger root already exists. "
+                            + "Are you sure to format bookkeeper metadata? "
+                            + "This may cause data loss.");
                 }
             }
 
-            // Clear the INSTANCEID
-            try {
-                zkc.delete(conf.getZkLedgersRootPath() + "/"
-                        + BookKeeperConstants.INSTANCEID, -1);
-            } catch (KeeperException.NoNodeException e) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("INSTANCEID not exists in zookeeper to delete");
-                }
+            if (!doFormat) {
+                return false;
             }
 
-            // create INSTANCEID
-            String instanceId = UUID.randomUUID().toString();
-            zkc.create(conf.getZkLedgersRootPath() + "/"
-                    + BookKeeperConstants.INSTANCEID, instanceId.getBytes(UTF_8),
-                    zkAcls, CreateMode.PERSISTENT);
+            BookKeeper bkc = new BookKeeper(new ClientConfiguration(conf));
+            bkc.ledgerManagerFactory.format(conf, bkc.regClient.getLayoutManager());
 
-            LOG.info("Successfully formatted BookKeeper metadata");
-        } finally {
-            if (null != bkc) {
-                bkc.close();
-            }
-            if (null != zkc) {
-                zkc.close();
-            }
+            return rm.format(conf);
```

Comment:
it doesn't look appropriate to make 3 different calls here

rm.prepareFormat(conf)
bkc.ledgerManagerFactory.format(conf, bkc.regClient.getLayoutManager())
rm.format(conf)

is it not possible to abstract it out to single rm method call, probably with arguments?

Reply:
thanks. ledger manager factory only manages ledger metadata, registration manager manages layout, cookies, instance id. so they are currently covering different metadata. so we need separate calls at this moment.
for end-user, they would still use the same `format` method, it doesn't change.

Reply:
@reddycharan since this is a code/interface refactor, if you feel we can combine some some of the logics into one, how about do this in a separate PR? I think is better to keep an interface refactor simple rather than changing too much logic.

Reply:
Hey @sijie, I'm working on having new set of bookieshellcommands - nukeexistingcluster and initnewcluster (by splitting metaformat. but leave metaformat as it is). Waiting for this commit to go in. Will consider refactoring while rebasing my commit on the new code, since it touches all these calls. Thanks.

Reply:
cool. thanks @reddycharan